ComponentKit/Utilities/CKComponentAction.h

```diff
@@ -11,44 +11,180 @@
 #import <UIKit/UIKit.h>
 
 #import <ComponentKit/CKComponentViewAttribute.h>
+#import <ComponentKit/CKComponentActionInternal.h>
 
 @class CKComponent;
 
 /**
- A component action is simply a selector. The selector may optionally take one argument: the sending component.
+ CKTypedComponentAction is a struct that represents a method invocation that can be passed to a child component to
+ trigger a method invocation on a target.
+ 
+ We allow a typed specification of the parameters that will be provided as arguments to the component action at runtime
+ through the variadic templated arguments. You may specify an arbitrary number of arguments to your component action,
+ and you may use either object, or primitive arguments. Only trivially-constructible arguments or pointers can be used
+ as types for component actions, so primitive types like int, CGRect, and NSObject * are fine. This is enforced via a
+ compile time check.
+
+ Methods will always be provided the sender as the first argument.
+ 
+ Usage in your component header:
+ 
+ @interface MyComponent : CKComponent
+ + (instancetype)newWithAction:(CKTypedComponentAction<NSString *, int>)action;
+ @end
+ 
+ When creating the action:
+ 
+ Option 1 - Scope action. Similar to target/selector action in that it skips the responder chain from the sender, and
+            directly invokes the selector on the component or controller corresponding with the scope. Promotion is also
```

Comment:
Yeah, I tried to make this clear, but what I mean by promotion is providing an action that receives less arguments than the declaration. So for instance, if my action is declared as:

```
- (instancetype)newWithAction:(CKTypedComponentAction<id, NSString *, int>)action
```

Then a *promotion* would be passing a typed action that receives less than the number of arguments in the declaration of the type.

So, these would all be promotions:

`[MyComponent newWithAction:CKTypedComponentAction<>(scope, @selector(mySelector))]`
`[MyComponent newWithAction:CKTypedComponentAction<id>(scope, @selector(mySelectorWithSender:firstArgument:))]`
`[MyComponent newWithAction:CKTypedComponentAction<id, NSString *>(scope, @selector(mySelectorWithSender:firstArgument:stringArgument:))]`

Any of the above would be implicitly converted to the type declared in the component header that receives the action.

**demotions** on the other hand, go from an action type that receives N arguments to a type that provides less than N arguments.  So for instance given this interface:

```
- (instancetype)newWithAction:(CKTypedComponentAction<id>)action
```

All of the following would be demotions:

`[MyComponent newWithAction:CKTypedComponentAction<id, NSString *>(scope, @selector(mySelectorWithSender:firstArgument:stringArgument))]`
`[MyComponent newWithAction:CKTypedComponentAction<id, int>(scope, @selector(mySelectorWithSender:firstArgument:intArgument))]`

These will fail to compile, since we're slicing off one of the arguments that the action expects. You can force a demotion through an explicit conversion constructor:

`[MyComponent newWithAction:CKTypedComponentAction<id>(CKTypedComponentAction<id, int>(scope, @selector(mySelectorWithSender:firstArgument:intArgument)))]`

But this should rarely be used. You can see above where I've had to do this for a button action that is being passed to the accessibility system. The accessibility API takes a CKTypedComponent<>, but in the button API, we've provided an action of type CKTypedComponent<UIEvent *>.

Anyway, in practice, the action is called, as it is currently with performSelector:withObject, except we use NSInvocation which will provide nil (or zeroed) arguments for any unspecified parameters, up to a certain size.

`[MyComponent newWithAction:CKTypedComponentAction<id>(scope, @selector(mySelectorWithSender:firstArgument:))]`
`[MyComponent newWithAction:CKTypedComponentAction<id, NSString *>(scope, @selector(mySelectorWithSender:firstArgument:stringArgument:))]`